src/main/java/org/sonar/plugins/csharp/SarifParser.java

```diff
@@ -44,35 +48,64 @@ public void parse(File file) {
+    if (root.has("runLogs")) {
+      JsonElement runLogs = parser.parse(contents).getAsJsonObject().get("runLogs");
+      for (JsonElement runLogElement: runLogs.getAsJsonArray()) {
+        JsonObject runLog = runLogElement.getAsJsonObject();
+        JsonArray results = runLog.getAsJsonArray("results");
+        if (results != null) {
+          handleIssues(results, false, true);
+        }
+      }
+    } else if (root.has("issues")) {
+      JsonElement issues = parser.parse(contents).getAsJsonObject().get("issues");
+      handleIssues(issues.getAsJsonArray(), true, false);
+    }
+  }
 
-        String ruleId = issue.get("ruleId").getAsString();
+  private void handleIssues(JsonArray issues, boolean linesOffByOne, boolean convertUri) {
+    for (JsonElement issueElement : issues) {
+      JsonObject issue = issueElement.getAsJsonObject();
+      handleIssue(issue, linesOffByOne, convertUri);
+    }
+  }
 
-        String message = issue.get(issue.has("shortMessage") ? "shortMessage" : "fullMessage").getAsString();
-        boolean hasLocation = false;
-        for (JsonElement locationElement : issue.get("locations").getAsJsonArray()) {
-          JsonObject location = locationElement.getAsJsonObject();
-          if (location.has("analysisTarget")) {
-            for (JsonElement analysisTargetElement : location.get("analysisTarget").getAsJsonArray()) {
-              hasLocation = true;
-              JsonObject analysisTarget = analysisTargetElement.getAsJsonObject();
-              String uri = analysisTarget.get("uri").getAsString();
-              JsonObject region = analysisTarget.get("region").getAsJsonObject();
-              int startLine = region.get("startLine").getAsInt();
+  private void handleIssue(JsonObject issue, boolean linesOffByOne, boolean convertUri) {
+    String ruleId = issue.get("ruleId").getAsString();
 
-              callback.onIssue(ruleId, uri, message, startLine);
-            }
-          }
-        }
+    String message = issue.get(issue.has("shortMessage") ? "shortMessage" : "fullMessage").getAsString();
+    boolean hasLocation = false;
+    for (JsonElement locationElement : issue.get("locations").getAsJsonArray()) {
+      JsonObject location = locationElement.getAsJsonObject();
+      if (location.has("analysisTarget")) {
+        for (JsonElement analysisTargetElement : location.get("analysisTarget").getAsJsonArray()) {
+          hasLocation = true;
+          JsonObject analysisTarget = analysisTargetElement.getAsJsonObject();
+
+          String uri = analysisTarget.get("uri").getAsString();
+          String absolutePath = convertUri ? uriToAbsolutePath(uri) : uri;
 
-        if (!hasLocation) {
-          callback.onProjectIssue(ruleId, message);
+          JsonObject region = analysisTarget.get("region").getAsJsonObject();
+          int startLine = region.get("startLine").getAsInt();
+          int line = linesOffByOne ? startLine + 1 : startLine;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS864)